wpiutil/src/main/native/cpp/FileLogger.cpp

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+#include "wpi/FileLogger.h"
+
+#ifdef __linux__
+#include <fcntl.h>
+#include <sys/inotify.h>
+#include <unistd.h>
+#endif
+
+#include <string>
+#include <string_view>
+#include <tuple>
+#include <utility>
+
+#include <fmt/format.h>
+
+#include "wpi/StringExtras.h"
+
+namespace wpi {
+FileLogger::FileLogger(std::string_view file,
+                       std::function<void(std::string_view)> callback)
+#ifdef __linux__
+    : m_fileHandle{open(file.data(), O_RDONLY)},
+      m_inotifyHandle{inotify_init()},
+      m_inotifyWatchHandle{
+          inotify_add_watch(m_inotifyHandle, file.data(), IN_MODIFY)},
+      m_thread{[=, this] {
+        char buf[4000];
+        struct inotify_event ev;
+        int len = 0;
+        while ((len = read(m_inotifyHandle, &ev, sizeof(ev))) > 0) {
+          int bufLen = 0;
+          if ((bufLen = read(m_fileHandle, buf, sizeof(buf)) > 0)) {
+            callback(std::string_view{buf, static_cast<size_t>(bufLen)});
+          }
+        }
+      }}
+#endif
+{
+}
+FileLogger::FileLogger(std::string_view file, log::DataLog& log,
+                       std::string_view key)
+    : FileLogger(file, [entry = log.Start(key, "string"),
+                        buf = wpi::SmallVector<char, 16>{},
+                        &log](std::string_view data) mutable {
+        if (!wpi::contains(data, "\n")) {
+          buf.append(data.begin(), data.end());
+          return;
+        }
+        std::string_view left;
+        std::string_view right;
+        std::string combinedData = fmt::format("{}{}", buf.data(), data);
+        buf.clear();
+        do {
+          std::tie(left, right) = wpi::rsplit(data, "\n");
```

Comment:
It doesn't seem correct to use [`rsplit()`](<https://github.wpilib.org/allwpilib/docs/development/cpp/namespacewpi.html#a7a31eb74242495e03e1745ab82085c81>) here, since it splits around the last occurrence, meaning that the newlines will be in `left`, not `right`.